### PR TITLE
Speedup scan grad() method

### PR DESCRIPTION
**Pull request by carriepl**
This is a second take for the PR #3435 which was merged by mistake and then subsequently reverted. It contains the same changes but the commits have been rebased together so they would not already be in the branch history and github would allow this PR.

@abergeron or @lamblin : please review this one instead of #3435

### CODE REVIEW

File: theano/scan_module/scan_op.py [hunk 1]
```diff
@@ -1938,39 +1938,45 @@ def get_out_idx(iidx):
...
-                   (connection_pattern[
-                       get_inp_idx(self_inputs.index(x))][odx])]
+        def compute_all_gradients(known_grads):
+            y_s = known_grads.keys()
+            g_y_s = known_grads.values()
+
+            for g_y in g_y_s:
+                if 'int' in str(g_y.dtype):
+                    raise TypeError("Gradients may never be integers but g_y "
+                            "has type " + str(g_y.type))
```
Comment (abergeron) [inline theano/scan_module/scan_op.py]: Flake8 here?

**Comment by abergeron** [comment]
I now did a review of the scan parts and I'm ok with a merge assuming travis passes.

**Comment by nouiz** [comment]
Just wait wednesday to merge. Just for safety.

File: theano/gradient.py [hunk 2]
```diff
@@ -1128,6 +1141,19 @@ def try_to_copy_if_needed(var):
             # we won't be able to post-process out the Nones if it does that
             input_grads = list(input_grads)
 
+            # Need to propagate the NullType gradients; if an input grad is
+            # not disconnected and the corresponding input is connected
+            # to at least one output whose gradient is NullType then the input
+            # grad should be NullType.
+            op_conn_pattern = _node_to_pattern(node)
```
Comment (lamblin) [inline theano/gradient.py]: No need to call `_node_to_pattern(node)` again, it should have already be called here: https://github.com/carriepl/Theano/blob/scan_grad_speedup2/theano/gradient.py#L985

**Comment by lamblin** [comment]
I had a look at the `grad` part, and I'm fine with it.

**Comment by carriepl** [comment]
Did both the flake8 issue @abergeron mentionned and the removed the useless function call that @lamblin identified. If Travis passes, then I guess all is left is waiting until Wednesday for a merge.

**Comment by carriepl** [comment]
Ok, travis passes with the changes.

**Comment by abergeron** [comment]
Unfortunately I've seen a bug in this with some code that David Kruger has.  This will crash in the grad() method with an IndexError.

I don't have any more details for now.

**Comment by caglar** [comment]
Yes that PR causes the IndexError as Arnaud mentioned. I got the error in
a  script where I you specify known_grads for a particular node in grad()
function. The full trace is as follows:

INFO:nmodel:Computing the gradients.
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/ipdb/__main__.py", line 157, in
main
    pdb._runscript(mainpyfile)
  File "/usr/lib/python2.7/pdb.py", line 1233, in _runscript
    self.run(statement)
  File "/usr/lib/python2.7/bdb.py", line 400, in run
    exec cmd in globals, locals
  File "<string>", line 1, in <module>
  File "train.py", line 1, in <module>
    from state import *
  File "train.py", line 12, in main
    exp_fn()
  File "state.py", line 2337, in
proto_ntm_fb_BABI_task_grucontroller_simple_small_10k_hard_task2_bowout_qmask
    prefix=prfx)
  File "mainloop.py", line 93, in __init__
    self.prepare_model()
  File
"/home/gulcehrc/Codes/python/MemoryNetwork/codes/caglar/core/basic.py",
line 138, in prepare_model
    mdl_name=mdl_name)
  File "nmodel.py", line 773, in get_train_fn
    grads = safe_grad(cost, params, known_grads=self.known_grads)
  File
"/home/gulcehrc/Codes/python/MemoryNetwork/codes/caglar/core/utils/utils.py",
line 212, in safe_grad
    grad_list = T.grad(cost, params, known_grads=known_grads,
add_names=True)
  File "/home/gulcehrc/libs/Theano/theano/gradient.py", line 567, in grad


On Wed, Sep 30, 2015 at 8:16 PM, abergeron <notifications@github.com> wrote:

> Unfortunately I've seen a bug in this with some code that David Kruger
> has. This will crash in the grad() method with an IndexError.
>
> I don't have any more details for now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/3461#issuecomment-144580025>.
>



-- 
Caglar GULCEHRE

**Comment by nouiz** [comment]
Can one of you give a way to reproduce this?

**Pull request merged**